### ISSUE TITLE
Cyclic dependency between twisted.python.reflect and twisted.python.deprecate

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4166]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4166|
|Type|defect|
|Created|2009-12-13 03:00:00Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-reflect-cycle-4166|

```
>>> from twisted.python import deprecate
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/exarkun/Projects/Twisted/branches/eliminate-zombie-warning-2078/twisted/python/deprecate.py", line 58, in <module>
    from twisted.python.reflect import fullyQualifiedName
  File "/home/exarkun/Projects/Twisted/branches/eliminate-zombie-warning-2078/twisted/python/reflect.py", line 817, in <module>
    from twisted.python.deprecate import deprecated
ImportError: cannot import name deprecated
>>> 

```

<details><summary>Searchable metadata</summary>

```
trac-id__4166 4166
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecate_reflect_cycle_4166 branches/deprecate-reflect-cycle-4166
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1260673200000000 1260673200000000
changetime__1266088443000000 1266088443000000
version__None None
owner__ 
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28163]) Branching to 'deprecate-reflect-cycle-4166'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Cycle broken

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_4"></a>|@jesstess set owner to @exarkun|
|-|-|

I don't feel qualified to make a value judgment on the mangling going on for ```_fullyQualifiedName```, but other than that:

 * [builds](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/clearer-keys-deprecations-4265) are happy
 * ```twisted/internet/error.py``` has a comment about this ticket that should go away or indicate that the problem is historical once this is closed. Actually, it looks like ```reflect``` isn't used at all in this file...
 * unit test of some kind?

Thanks for all your work!

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_5"></a>|@jesstess commented|
|-|-|

Ahem...build results [here](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-reflect-cycle-4166), actually. And I didn't actually run them. *sigh*

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

I deleted the comment and `reflect` import in `twisted/internet/error.py`.  (`reflect` was "unused" there, but that was intentional; the import was for the side-effect it created, of making the `twisted.python.deprecate` import which followed succeed.)  Thanks for catching that. :)

I don't think a unit test makes sense here.  The behavior is a result of the interaction between two (somewhat large) chunks of code which cannot be broken into isolated chunks easily (or at all, really, I think.)  _Some_ kind of automated test clearly makes sense, though (since automated testing is the best thing since sliced bread).  The test that seems to make sense to me is one that looks over a specified set of code (for example, all of Twisted) and marks import cycles as errors.

I looked around a bit for an existing such tool, but didn't have much luck.  I did find <http://svn.zope.org/Zope3/trunk/src/zope/importtool/?rev=25227> but it has since been deleted as far as I can tell.

Considering the difficulty of introducing an import cycle (you really know that you're doing it, since you suddenly have an import at the bottom of a file instead of the top...), I'm not _too_ worried about continuing to leave this behavior without automated test coverage.  It _would_ be nice to have someday, though.

As far as the `__name__` and `__module__` mangling goes, I had considered that and then decided against it (just barely) until I ran the unit tests, which actually rely on `fullyQualifiedName` (specifically) having a particular fully qualified name.

Thanks for the review.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_7"></a>|@jesstess set owner to @exarkun|
|-|-|

twisted/web/error.py needs a copyright bump. Other than that, looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [28196]) Merge deprecate-reflect-cycle-4166

Author: exarkun
Reviewer: jesstess
Fixes: [#4166](https://github.com/twisted/twisted/issues/4166)

Remove the cyclic dependency between `twisted.python.deprecate` and
`twisted.python.reflect` by moving the implementation of `fullyQualifiedName`
into the former.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
